File: bridges/voiceover/Sources/VoiceOverBridgeAdapters/Ports/Listener.swift
// ROLE: adapter seam for the accepting edge, a bound endpoint that yields one connection at a time.
// USED BY: BridgeServer.
// IMPLEMENTED BY: LocalSocketListener, TCPListener and FakeListener.

public struct ListenerClosed: Error {
	public init() {}
}

public protocol Listener: AnyObject {
	/// Human-readable address, a socket path or `host:port`, defined once `open()` has bound.
	var endpoint: String { get }

	/// Throws a bind failure on the caller's thread, so it reaches whoever asked for the bridge.
	func open() throws

	/// Throws `PollTimeout` when idle, and `ListenerClosed` once `close()` has been called.
	func accept() throws -> any Transport

	/// Stop listening. Idempotent, and it unblocks a pending `accept`.
	func close()
}

File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/adapters/ports/listener.py
# nvdaMcpBridge adapters -- the Listener seam.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: adapter seam for the accepting edge.
# USED BY: adapters/bridge_server.py.
# IMPLEMENTED BY: adapters/tcp_listener.py, adapters/named_pipe_listener.py and tests/fakes/listener.py.

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

if TYPE_CHECKING:
	from .transport import Transport


class ListenerClosed(Exception):
	"""Raised by accept() once the listener has been closed."""


class Listener(ABC):
	@property
	@abstractmethod
	def endpoint(self) -> str:
		"""Human-readable accepting address, defined once open() has bound."""

	@abstractmethod
	def open(self) -> None:
		"""Raises on a bind failure."""

	@abstractmethod
	def accept(self) -> Transport:
		"""TimeoutError after a poll window so the loop can check for a stop; ListenerClosed after close()."""

	@abstractmethod
	def close(self) -> None:
		"""Idempotent; unblocks a pending accept()."""
